### ISSUE TITLE
[web] Controls not working properly for non-US keyboards

### ISSUE THREAD

**Issue by thaliel** [ZQuestClassic/ZQuestClassic#1006]
I can use z on any quest exactly once. After that, pressing z does absolutely nothing

**Comment by connorjclark** [comment]
what version of ZC, what quest, what OS?

maybe a gamepad will work better for you?

**Comment by thaliel** [comment]
I (try to) play the nrwoser version on my windows laptop. Unfortunately, I do not have a functioning gamepad atm

**Comment by Jambu2024** [comment]
have you tried with multiple keyboards also which keyboard standard are you using

**Comment by connorjclark** [comment]
> I (try to) play the nrwoser version on my windows laptop. Unfortunately, I do not have a functioning gamepad atm

Did you mean "newest"? What version exactly?

Are you getting it from https://zquestclassic.com/releases/ ?

this is a strange issue that we haven't heard of before

**Comment by inventionate** [comment]
I think I can shed some light on this. I have the same problem. It seems to be related to choosing a non-American keyboard layout. If you have a German layout, for example, then the Z key is Y and that seems to cause confusion. A quick solution is to map A to a key other than Z or Y. It's not great and I hope this can be fixed with the info, but at least you can play this way (which is the most important thing). And in general, thank you very much for the great engine and fantastic web version!

**Comment by inventionate** [comment]
Is there any progress here? Unfortunately, the bug still seems to be present.

**Comment by connorjclark** [comment]
No update, but I'll prioritize it for the 3.0 release.

@inventionate can you confirm what browser you are using (probably doesn't matter but good to know). Also, if you could it'd be useful to know if you have the same issue for the native build – I'm pretty sure we squashed this bug there already. Please give the latest 3.0 prerelease a try if you can and let me know how it goes (+ what OS you are on).

**Comment by inventionate** [comment]
Thank you very much! I tested it with Safari (macOS 26.1). Since there is no nightly build for macOS available for download, I haven't been able to test it with the native build yet. It would be great if that could be resolved.

**Comment by connorjclark** [comment]
Oh that's just a hiccup with the latest release. The one just before has a Mac asset: https://github.com/ZQuestClassic/ZQuestClassic/releases/tag/3.0.0-prerelease.146%2B2025-12-04

**Comment by inventionate** [comment]
Thanks, I hope the hiccups go away quickly ;-) I installed the latest build and what can I say, it works great. Thank you very much! I hope this version will soon be available for the web version so that I can test/use it there. As I said, thank you!

**Comment by connorjclark** [comment]
Oh, so you're saying that the defaults are bad - that on a german keyboard what the browser version maps to the A/B buttons are not next to each other like they would be on a US keyboard. But once manually mapped it's otherwise all OK, right?

**Comment by inventionate** [comment]
Yes, it's a bit awkward on German keyboards because Z and Y are swapped. But the real deal breaker is that pressing Z on a German keyboard (in the current web version) only works once. After that, the key no longer works. However, if I manually reassign A and B, the problem doesn't occur. I suspect it has to do with the different keyboard layout. It doesn't happen in the current Nightly version either, which means I can press the Y key as often as I want and everything works fine.

**Comment by connorjclark** [comment]
Yup, managed to see it myself by changing my Mac's keyboard layout. Ran into the "only once" bug (it allows another german Y input if you hit german Z...). I found the reason for it, and it's a miracle anything ever worked at all. I also will get the Controls dialog to print the correct keyboard letter (rather than the name of the key in that position for a standard US keyboard), and also make the "Default controls" text in About reflect the keyboard layout too.

Will polish up the fix and try to deploy it tomorrow.

**Comment by inventionate** [comment]
Thank you very much :-)

**Comment by connorjclark** [comment]
Should be fixed as of https://github.com/ZQuestClassic/ZQuestClassic/commit/4bc18457b2e12561625f427136ffcda01b7a01fe. Let me know if it works now.

**Comment by inventionate** [comment]
Works wonderfully, thank you very much!
